pytorch_lightning/callbacks/model_checkpoint.py

```diff
@@ -0,0 +1,205 @@
+            to overwrite the current save file is made
+            based on either the maximization or the
+            minimization of the monitored quantity. For `val_acc`,
+            this should be `max`, for `val_loss` this should
+            be `min`, etc. In `auto` mode, the direction is
+            automatically inferred from the name of the monitored quantity.
+        save_weights_only (bool): if True, then only the model's weights will be
+            saved (`model.save_weights(filepath)`), else the full model
+            is saved (`model.save(filepath)`).
+        period (int): Interval (number of epochs) between checkpoints.
+
+    Example::
+
+        from pytorch_lightning import Trainer
+        from pytorch_lightning.callbacks import ModelCheckpoint
+
+        checkpoint_callback = ModelCheckpoint(filepath='my_path')
+        Trainer(checkpoint_callback=checkpoint_callback)
+
+        # saves checkpoints to my_path whenever 'val_loss' has a new min
+    """
+
+    def __init__(self, filepath, monitor='val_loss', verbose=0,
+                 save_top_k=1, save_weights_only=False,
+                 mode='auto', period=1, prefix=''):
+        super(ModelCheckpoint, self).__init__()
+        if (
+            save_top_k and
+            os.path.isdir(filepath) and
+            len(os.listdir(filepath)) > 0
+        ):
+            warnings.warn(
+                f"Checkpoint directory {filepath} exists and is not empty with save_top_k != 0."
+                "All files in this directory will be deleted when a checkpoint is saved!"
+            )
+
+        self.monitor = monitor
+        self.verbose = verbose
+        self.filepath = filepath
+        os.makedirs(filepath, exist_ok=True)
+        self.save_top_k = save_top_k
+        self.save_weights_only = save_weights_only
+        self.period = period
+        self.epochs_since_last_check = 0
+        self.prefix = prefix
+        self.best_k_models = {}
+        # {filename: monitor}
+        self.kth_best_model = ''
+        self.best = 0
+        self.save_function = None
+
+        if mode not in ['auto', 'min', 'max']:
+            warnings.warn(
+                f'ModelCheckpoint mode {mode} is unknown, '
+                'fallback to auto mode.', RuntimeWarning)
+            mode = 'auto'
+
+        if mode == 'min':
+            self.monitor_op = np.less
+            self.kth_value = np.Inf
+            self.mode = 'min'
+        elif mode == 'max':
+            self.monitor_op = np.greater
+            self.kth_value = -np.Inf
+            self.mode = 'max'
+        else:
+            if 'acc' in self.monitor or self.monitor.startswith('fmeasure'):
+                self.monitor_op = np.greater
+                self.kth_value = -np.Inf
+                self.mode = 'max'
+            else:
+                self.monitor_op = np.less
+                self.kth_value = np.Inf
+                self.mode = 'min'
+
+    def _del_model(self, filepath):
```

Comment:
this is quite strange, why to create the whole dir tree and then delete it?

Reply:
A test fails without this. So I'll let it for now.